src/chan_vese.jl

```diff
@@ -0,0 +1,252 @@
+using MosaicViews
+using LazyArrays
+using BenchmarkTools
+using Images, TestImages
+using ImageBase.ImageCore: GenericGrayImage, GenericImage
+
+function calculate_averages(img::AbstractArray{T, N}, H𝚽::AbstractArray{T, M}) where {T<:Real, N, M}
+    H𝚽ⁱ = @. 1. - H𝚽
+    ∫H𝚽 = sum(H𝚽)
+    ∫H𝚽ⁱ = sum(H𝚽ⁱ)
+    if ndims(img) == 2
+        ∫u₀H𝚽 = sum(img .* H𝚽)
+        ∫u₀H𝚽ⁱ = sum(img .* H𝚽ⁱ)
+    elseif ndims(img) == 3
+        ∫u₀H𝚽 = sum(img .* H𝚽, dims=(1, 2))
+        ∫u₀H𝚽ⁱ = sum(img .* H𝚽ⁱ, dims=(1, 2))
+    end
+    if ∫H𝚽 != 0
+        c₁ = ∫u₀H𝚽 / ∫H𝚽
+    end
+    if ∫H𝚽ⁱ != 0
+        c₂ = ∫u₀H𝚽ⁱ / ∫H𝚽ⁱ
+    end
+
+    return c₁, c₂
+end
+
+function difference_from_average_term(img::AbstractArray{T, N}, H𝚽::AbstractArray{T, M}, λ₁::Float64, λ₂::Float64) where {T<:Real, N, M}
+    c₁, c₂ = calculate_averages(img, H𝚽)
+
+    if ndims(img) == 2
+        return @. -λ₁ * (img - c₁)^2 + λ₂ * (img - c₂)^2
+    elseif ndims(img) == 3
+        return -λ₁ .* sum((img .- c₁).^2, dims=3) .+ λ₂ .* sum((img .- c₂).^2, dims=3)
+    end
+end
+# H𝚽 = LazyArray(@~ @. 1. * (𝚽ⁿ > 0))
+function _calculate_averages(img::AbstractArray{T, N}, 𝚽ⁿ::AbstractArray{T, M}) where {T<:Real, N, M}
+    ∫H𝚽 = ∫H𝚽ⁱ = ∫u₀H𝚽 = ∫u₀H𝚽ⁱ = 0
+
+    for i in CartesianIndices(img)
+        H𝚽 = 1. * (𝚽ⁿ[i] > 0)
+        H𝚽ⁱ = 1. - H𝚽
+        ∫H𝚽 += H𝚽
+        ∫H𝚽ⁱ += H𝚽ⁱ
+        ∫u₀H𝚽 += img[i] * H𝚽
+        ∫u₀H𝚽ⁱ += img[i] * H𝚽ⁱ
+    end
+    if ∫H𝚽 != 0
+        c₁ = ∫u₀H𝚽 ./ ∫H𝚽
+    end
+    if ∫H𝚽ⁱ != 0
+        c₂ = ∫u₀H𝚽ⁱ ./ ∫H𝚽ⁱ
+    end
+
+    return c₁, c₂
+end
+
+function δₕ(x::AbstractArray{T,N}, h::Float64=1.0) where {T<:Real, N}
+    return @~ @. h / (h^2 + x^2)
+end
+
+function initial_level_set(shape::Tuple)
+    x₀ = reshape(collect(0:shape[begin]-1), shape[begin], 1)
+    y₀ = reshape(collect(0:shape[begin+1]-1), 1, shape[begin+1])
+    𝚽₀ = @. sin(pi / 5 * x₀) * sin(pi / 5 * y₀)
+end
+
+function chan_vese(img::GenericGrayImage;
+                    μ::Float64=0.25,
+                    λ₁::Float64=1.0,
+                    λ₂::Float64=1.0,
+                    tol::Float64=1e-3,
+                    max_iter::Int64=500,
+                    Δt::Float64=0.5,
+                    reinitial_flag::Bool=false) #where {T<:Real, N}
+    img = float64.(channelview(img))
+    iter = 0
+    h = 1.0
+    m, n = size(img)
+    s = m * n
+    𝚽ⁿ = initial_level_set((m, n)) # size: m * n
+    del = tol + 1
+    img .= img .- minimum(img)
+
+    if maximum(img) != 0
+        img .= img ./ maximum(img)
+    end
```

Comment:
Why is this standardization needed?

Reply:
About standardization, I think a standardized image may contribute to a segmentation algorithm since it makes the gray levels of image more dispersive. If the grey levels of image range from 0.2 to 0.8, it seems better to standardize image so that the grey levels will range from 0.0 to 1.0.
The test of `chan_vese` use a resized `64*64 cameraman` as the test image, whose grey levels range from 0.02 to 0.988. The standardization has showed that it will cause some difference :
```julia
julia> using TestImages, Images

julia> img_512 = testimage("cameraman");

julia> maximum(img_512)
Gray{N0f8}(1.0)

julia> minimum(img_512)
Gray{N0f8}(0.0)

julia> img_64 = imresize(testimage("cameraman"), (64, 64));

julia> maximum(img)
Gray{N0f8}(0.988)

julia> minimum(img)
Gray{N0f8}(0.02)
```
Test image shows as following:
![cameraman_64](https://user-images.githubusercontent.com/73411354/134764230-993dbd2f-2ee2-4610-9326-83fc55194bf1.png)

```julia
julia> res_no_std = chan_vese(img, μ=0.1, tol=1e-2, max_iter=200);

julia> sum(res_no_std)
1075

julia> colorview(Gray, res_no_std)

```
![res_64_no_std](https://user-images.githubusercontent.com/73411354/134764113-470b922e-1571-4720-9493-c3f57fa7bab1.png)

```julia
julia> res_std = chan_vese(img, μ=0.1, tol=1e-2, max_iter=200);

julia> sum(res_std)
1064

julia> colorview(Gray, res_std)

```
![res_64_std](https://user-images.githubusercontent.com/73411354/134764120-baadaeef-4b20-4eac-91ab-a227096ef601.png)

We can find that there are some difference. Without standardization, the part between the man's leg can't be segmented well. So maybe we have to maintain the standardization?

Reply:
Maybe we can introduce a keyword `normalize=false` to control this behavior. And we can also refer to ImageContrastAdjustment in the docstring for more complicated normalization workflows.

Reply:
Presumably that is easily compensated by changing the values of `λ`. The only place in the mathematics where the image magnitude means anything is the `λ*sum(abs2, img[componentmask] .- componentmean)`. So anything you can do by standardizing the image, you can do more quickly by adjusting `λ`.

What does standardization even mean for an image with RGB pixels? Do you take the min & max across channels too? Why does the magnitude of the blue channel scale the magnitude of the red one? Aren't these independent pieces of information?